tlair/src/test/scala/at/forsyte/apalache/tla/typecomp/pbt/TypecompPBT.scala

```diff
@@ -0,0 +1,123 @@
+            Gen.nonEmptyListOf(genInstr(t)).map(elems => builder.seq(elems: _*))
+          case t: TupT1 =>
+            for {
+              args <- Gen.sequence[Seq[T], T](t.elems.map(genInstr))
+            } yield builder.tuple(args: _*)
+          // `None` because genTlaType1 only generates closed row types
+          case RecRowT1(RowT1(fieldTypes, None)) =>
+            fieldTypes.map { case (field, typ) => genInstr(typ).map(v => (field, v)) } match {
+              case Nil    => Gen.fail // No possible value for the empty record
+              case fields => Gen.sequence[Seq[(String, T)], (String, T)](fields).map(fs => builder.rowRec(None, fs: _*))
+            }
+          // `None` because genTlaType1 only generates closed row types
+          case t @ VariantT1(RowT1(fieldTypes, None)) =>
+            fieldTypes.map { case (tag, typ) =>
+              genInstr(typ).map(arg => builder.variant(tag, arg, t))
+            }.toList match {
+              case Nil            => Gen.fail // No possible value for the empty variant
+              case v :: Nil       => v // Just one variant option
+              case v0 :: v1 :: vs => Gen.oneOf(v0, v1, vs: _*)
+            }
+          case FunT1(argT, resT) =>
+            for {
+              vName <- genName(argT)
+              dom <- genSet(argT)
+              res <- genInstr(resT)
+            } yield builder.funDef(res, (vName, dom))
+          case VariantT1(RowT1(_, Some(_))) | RecRowT1(RowT1(_, Some(_))) =>
+            // Does not correspond to a concrete term
+            // TODO: In future can generate arbitrary rows to fill in the vars here
+            Gen.fail
+          case RealT1 | _: OperT1 | _: VarT1 | _: RecT1 | _: RowT1 | _: SparseTupT1 =>
+            // TlaType1ConcreteGen should ensure these these types should never be createc
+            throw new IllegalArgumentException(
+                s"Type ${typ.getClass().getSimpleName()} is not supported in PBT generation but given type ${typ}")
+        }
+      }
+
+      genInstr(typ)
+    }
+
+    lazy val instr: Gen[T] = for {
+      typ <- contreteTlaType1
+      inst <- ofType(typ)
+    } yield inst
+
+    val typeVar: Gen[String] = Gen.alphaLowerChar.map(_.toString)
+
+    val rowVar = Gen.some(typeVar)
+    lazy val genRecordFields = for {
+      // We need to generate unique names for record fields
+      fieldNames <- Gen.nonEmptyContainerOf[Set, String](Gen.identifier)
+      fields <- Gen.sequence[Seq[(String, TBuilderInstruction)], (String, TBuilderInstruction)](
+          fieldNames.map(n => Gen.zip(Gen.const(n), instr)))
+    } yield fields
+  }
+
+  /** Implicit arbitrary instances for [TBuilderInstruction]s */
+  object ArbTBuilderInstruction {
+    implicit val argContreteTlaType1: Arbitrary[TlaType1] = Arbitrary(GenTBuilderInstruction.contreteTlaType1)
+    implicit val arbTBuilderInstruction: Arbitrary[TBuilderInstruction] = Arbitrary(GenTBuilderInstruction.instr)
+  }
```

Comment:
Why the separate object here?
It looks like scope control that could be done via access modifiers?

Reply:
I've added a comment to clarify: 6bd5d3231261ec9994af5373c830ec03a081a1d5